037-prototypes/proposal.md

```diff
@@ -0,0 +1,467 @@
+
+```go
+// Object is an object receiving messages and being emitted in response to
+// messages.
+type Object map[string]interface{}
+
+// InfoRequest is the payload written to stdin for the `./info` script.
+type InfoRequest struct {
+  // The properties specifying the object to act on.
+  Object Object `json:"object"`
+
+  // Configuration for handling TLS.
+  TLS TLSConfig `json:"tls,omitempty"`
+}
+
+// InfoResponse is the payload written to stdout from the `./info` script.
+type InfoResponse struct {
+  // The version of the prototype interface that this prototype conforms to.
+  InterfaceVersion string `json:"interface_version"`
+
+  // An optional icon to show to the user.
+  //
+  // Icons must be namespaced by in order to explicitly reference an icon set
+  // supported by Concourse, e.g. 'mdi:' for Material Design Icons.
+  Icon string `json:"icon,omitempty"`
+
+  // The messages supported by the object.
+  Messages []string `json:"messages,omitempty"`
+}
+
+// MessageRequest is the payload written to stdin for a message.
+type MessageRequest struct {
+  // The properties specifying the object to act on.
+  Object Object `json:"object"`
+
+  // Configuration for handling TLS.
+  TLS TLSConfig `json:"tls,omitempty"`
+
+  // Path to a file into which the message handler must write its response.
+  ResponsePath string `json:"response_path"`
+}
+
+// MessageResponse is written to the `response_path` for each object returned
+// by the message. Multiple responses may be written to the same file,
+// concatenated as a JSON stream.
+type MessageResponse struct {
+  // The identifying properties for the object.
+  Object Object `json:"object"`
+
+  // Metadata to associate with the properties. Shown to the user.
+  Metadata []MetadataField `json:"metadata,omitempty"`
+}
+
+// TLSConfig captures common configuration for communicating with servers over
+// TLS.
+type TLSConfig struct {
+  // An array of CA certificates to trust.
+  CACerts []string `json:"ca_certs,omitempty"`
+
+  // Skip certificate verification, effectively making communication insecure.
+  SkipVerification bool `json:"skip_verification,omitempty"`
+}
+
+// MetadataField represents a named bit of metadata associated to an object.
+type MetadataField struct {
+  Name  string `json:"name"`
+  Value string `json:"value"`
+}
+```
+
+
+## Prototype Info
+
+Prior to sending any message, the default command (i.e.
+[`CMD`](https://docs.docker.com/engine/reference/builder/#cmd)) for the image
+will be executed with an `InfoRequest` piped to `stdin`. This request contains
+an **object**.
+
+The command must write an `InfoResponse` to `stdout` in response. This response
+specifies the prototype interface version that the prototype conforms to, an
+optional icon to show in the UI, and the messages supported by the given
+object.
+
+### Example **info** request/response
+
+Request sent to `stdin`:
+
+```json
+{ "object": { "uri": "https://github.com/concourse/concourse" } }
```

Comment:
It's unclear to me where the `uri` field is coming from (specifically for an `InfoRequest`)? If there's any such field to be passed, wouldn't it be something more akin to `ATC_EXTERNAL_URL`? What about the other environment variables that Concourse sets today for `get/put` messages - why are these messages special in that they would receive environment variables that other messages don't receive?

Reply:
The `object` field is a black box object. In this case it's similar to the `source:` of a `git` resource in a pipeline. It is not the ATC external URL. I'll try to clarify this in the doc. Thanks!

Reply:
Regarding `ATC_EXTERNAL_URL` and friends - right now we propagate that to resources in service of notification resources. So I think we can integrate them into a new RFC outlining how prototypes can be used for emitting notifications.

Instead of environment variables, they would probably be injected into the object. This would be done automatically by Concourse whenever a notification is emitted. So notification prototypes could receive an object kind of like this:

```json
{
  "webhook_uri": "https://hooks.slack.com/services/XXXX",
  "notification": {
    "cluster": {
      "url": "https://ci.example.com",
      "name": "my-cluster"
    },
    "build": {
      "id": 1234,
      "status": "succeeded",
    }
  }
}
```

The `notification` field here would be automatically injected into the object by Concourse. The prototype would then read any/all information it needs from it and we can retire the `ATC_*` env vars.